File: js/heap_sort.js
class Heap {
    constructor() {
        this.list = new ArrayList();
    }

    add(value) {
        this.list.push(value);

        let currentIndex = this.length - 1;

        while (currentIndex > 0) {
            let parentIndex = Math.floor((currentIndex - 1) / 2);

            if (this.compare(currentIndex, parentIndex) > 0) {
                let temp = this.list.get(currentIndex);
                this.list.set(currentIndex, this.list.get(parentIndex));
                this.list.set(parentIndex, temp);
            } else {
                break;
            }

            currentIndex = parentIndex;
        }
    }

    remove() {
        if (this.list.size() == 0) return null;

        let removedValue = this.list.get(0);

        this.list.set(0, this.list.get(this.list.size() - 1));
        this.list.remove();

        let currentIndex = 0;
        while (currentIndex < this.list.size()) {
            let leftChildIndex = 2 * currentIndex + 1;
            let rightChildIndex = 2 * currentIndex + 2;

            if (leftChildIndex >= this.list.size()) break;

            let maxIndex = leftChildIndex;
            if (rightChildIndex < this.list.size()) {
                if (this.compare(maxIndex, rightChildIndex) < 0) {
                    maxIndex = rightChildIndex;
                }
            }

            if (this.compare(currentIndex, maxIndex) < 0) {
                let temp = this.list.get(maxIndex);
                this.list.set(maxIndex, this.list.get(currentIndex));
                this.list.set(currentIndex, temp);
                currentIndex = maxIndex;
            } else {
                break;
            }
        }

        return removedValue;
    }
    compare(v1, v2) {
        return this.list.get(v1) - this.list.get(v2);
    }

    size() {
        return this.list.size();
    }
}

class ArrayList extends Array {
    get(index) {
        return this[index];
    }

    add(value) {
        this.push(value);
    }

    remove() {
        this.pop();
    }

    set(index, value) {
        this[index] = value;
    }

    size() {
        return this.length;
    }
}

function heapSort(list) {
    const heap = new Heap();

    for (let i = 0; i < list.length; i++) {
        heap.add(list[i]);
    }

    for (let i = list.length - 1; i >= 0; i--) {
        list[i] = heap.remove();
    }
}

function testHeapSort() {
    output.innerHTML += "<h3>testing heap sort in javascript...</h3>";
    const unsortedArray = getRandArray(10, 1);
    output.innerHTML += `<p>unsorted array -> [${unsortedArray}]</p>`;
    heapSort(unsortedArray);
    output.innerHTML += `<p>sorted array -> [${unsortedArray}]</p>`;
}

testHeapSort();

File: python/heap_sort.py
class Heap:
    def __init__(self):
        self.list = ArrayList()


    def add(self, value):
        self.list.add(value)

        currentIndex = self.list.size() - 1

        while currentIndex > 0:
            parentIndex = (currentIndex - 1) // 2

            if self.compare(currentIndex, parentIndex) > 0:
                temp = self.list.get(currentIndex)
                self.list.set(currentIndex, self.list.get(parentIndex))
                self.list.set(parentIndex, temp)
            else:
                break

            currentIndex = parentIndex


    def remove(self):
        if self.list.size() == 0:
            return None
        
        removedValue = self.list.get(0)

        self.list.set(0, self.list.get(self.list.size() - 1))
        self.list.remove()

        currentIndex = 0
        while currentIndex < self.list.size():
            leftChildIndex = 2 * currentIndex + 1
            rightChildIndex = 2 * currentIndex + 2

            if leftChildIndex >= self.list.size():
                break
        
            maxIndex = leftChildIndex
            if rightChildIndex < self.list.size():
                if self.compare(maxIndex, rightChildIndex) < 0:
                    maxIndex = rightChildIndex

            if self.compare(currentIndex, maxIndex) < 0:
                temp = self.list.get(maxIndex)
                self.list.set(maxIndex, self.list.get(currentIndex))
                self.list.set(currentIndex, temp)
                currentIndex = maxIndex
            else:
                break

        return removedValue


    def compare(self, v1, v2):
        return self.list.get(v1) - self.list.get(v2)
    

class ArrayList(list):
    def add(self, value):
        self.append(value)


    def remove(self):
        self.pop()


    def get(self, index):
        return self[index]
    

    def set(self, index, value):
        self[index] = value
    

    def size(self):
        return len(self)
    

def heap_sort(array):
    heap = Heap()

    for i in range(len(array)):
        heap.add(array[i])

    for i in range(len(array) - 1, -1, -1):
        array[i] = heap.remove()


print("testing heap sort in python...")

unsortedArray = [4, 1, 5, 2, 8, 3, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray))
heap_sort(unsortedArray)
print("sorted array -> " + str(unsortedArray))
